### PR TITLE
release-23.2: sqlstats: fix memory leak when writing statement insights and no transaction insights

### PR DESCRIPTION
**Pull request by xinhaoz**
Backport 1/1 commits from #122128

Backport 3/3 commits from #122699

/cc @cockroachdb/release

Release justification: bug fix

----

**sqlstats: delete sqlstats.StatsCollector interface**

There is one implementation of this interface. Let us
simply use the concrete type.

Epic: none

Release note: None

**sqlstats: introduce insights writer to StatsCollector**

This commit adds the insights writer to the `StatsCollector` struct. This
is in an effort to shift the responsibility of writing insights from being
owned by the `ssmemstorage.Container` to `sslocal.StatsCollector`. It makes
more sense for the StatsCollector, which is already responsible for
coordinating the writing of sql stats for a conn executor, to also
orchestrate insights collection operations. The sql stats application
container's intended purpose is to store per-application sql statistics.

Epic: none
Release note: None

**sqlstats,insights: use StatsCollector to record stmt insights**

This commit moves the recording of statement insights one layer up.
This decouples the writing of statement insights from the writing
of statement statistics to the sql stats containers. Previously,
if writing statement stats was throttled we also skipped writing
statement insights. The two processes may now occur independently.
Additionally, we introduce the flag `sendInsights` to the stats
collector, which is set at the start of every transaction. This
flag will be used to determine whether to observe insights for the
entire transaction. This ensures that writing statement insights
-> writing transaction insights so that the statement insight memory
will be freed.

Epic: none
Release note: None

**sqlstats,insights: use StatsCollector to record txn insights**

This commit moves the recording of transaction insights one layer up.
This decouples the writing of transaction insights from the writing
of transaction statistics to the sql stats containers. Previously,
if writing transaction stats was throttled we also skipped writing
transaction insights. This can lead to a memory leak if we've already
recorded statement insights for this transaction, as that memory is only
released once the corresponding transaction insight is sent. The two
processes may now occur independently.

This fixes a memory leak in the scenario where we continue to send
statement insights to the insights ingester without sending any
transaction insights. The ingester frees the memory used by statement
insights once it receives a transaction event. Previously, due to the
coupling of writing sql stats and insights, if we transaction stats
were throttled (e.g. due to in-memory capacities), we would start
accumulating statement insights.

Epic: none
Fixes: https://github.com/cockroachdb/cockroach/issues/121803
Release note: None

### CODE REVIEW

File: pkg/sql/executor_statement_metrics.go [hunk 1]
```diff
@@ -255,6 +255,10 @@ func (ex *connExecutor) recordStatementSummary(
 		}
 	}
 
+	if stmtFingerprintID != 0 {
+		ex.statsCollector.ObserveStatement(stmtFingerprintID, recordedStmtStats)
```
Comment (dhartunian) [inline pkg/sql/executor_statement_metrics.go:259]: In retrospect, should we rename `ObserveStatement` to be more insights-specific? It doesn't affect other statement stuff right?

I think it's confusing with `RecordStatement` above.

Comment (dhartunian) [reply pkg/sql/executor_statement_metrics.go:259]: Do we need the conditional here, can't the method handle the 0 ID?

Comment (dhartunian) [reply pkg/sql/executor_statement_metrics.go:259]: Additionally, why can't `RecordStatement` call `ObserveStatement`?

File: pkg/sql/conn_executor_exec.go [hunk 2]
```diff
@@ -3320,6 +3320,8 @@ func (ex *connExecutor) recordTransactionFinish(
 		)
 	}
 
+	ex.statsCollector.ObserveTransaction(ctx, transactionFingerprintID, recordedTxnStats)
```
Comment (dhartunian) [inline pkg/sql/conn_executor_exec.go:3323]: Similar to prior commit, do we need a conditional here to check if ID is 0?

File: pkg/sql/executor_statement_metrics.go (hunk 1 above)
Comment (xinhaoz) [reply pkg/sql/executor_statement_metrics.go:259]: This PR aims to separate insights and sql stats writing and we accomplished this by moving ObserveStatement to this layer. These ops can occur independently whereas they were being coupled before (with insights being scrapped if we don't record sql stats).

File: pkg/sql/conn_executor_exec.go (hunk 2 above)
Comment (xinhaoz) [reply pkg/sql/conn_executor_exec.go:3323]: I'll remove the stmtFIngerprintId == 0 check in the statement recording. It's not necessary.

**Review by dhartunian — APPROVED** [review]

**Review by rafiss — APPROVED** [review]

**Pull request merged**